src/daemon/auth/providers/openid_connect/provider.rs

```diff
@@ -587,20 +616,61 @@ impl AuthProvider for OpenIDConnectAuthProvider {
+                    return Ok(Some(ActorDef::user(session.id, session.attributes, None)));
                 }
 
-                Ok(Some(ActorDef::user(session.id, session.attributes, new_auth)))
+                let new_auth = match self.try_refresh_token(&session) {
+                    Ok(auth) => {
+                        trace!(
+                            "OpenID Connect: Succesfully refreshed token for user \"{}\"",
+                            &session.id
+                        );
+                        auth
+                    }
+                    Err(err) => {
+                        trace!("OpenID Connect: rfc-6749 5.2 Error Response returned...");
+                        debug!(
+                            "OpenID Connect: Refreshing the token for user '{}' failed: {}",
+                            &session.id, &err
+                        );
+                        match err {
+                            // This is the Error returned by the OpenID Connect Provider if the session was terminated
+                            // by them. The user should be able to create a new session by logging in again.
+                            CoreErrorResponseType::InvalidGrant => {
+                                warn!("OpenID Connect: invalid_grant {:?}", err);
+                                return Err(Error::ApiInvalidCredentials("The Session has ended.".to_string()));
+                            }
+                            CoreErrorResponseType::InvalidRequest | CoreErrorResponseType::InvalidClient => {
+                                warn!("OpenID Connect:  rfc-6749 5.2 {:?}", err);
+                                return Err(Error::ApiAuthPermanentError(
+                                    "There is a unknown Authentication Problem.".to_string(),
+                                ));
+                            }
+                            // If changes are made to the roles of the user, the client or the scope on the side of the OpenID Connect Provider,
+                            // the token refresh may get one of these errors.
+                            CoreErrorResponseType::UnauthorizedClient
+                            | CoreErrorResponseType::UnsupportedGrantType
+                            | CoreErrorResponseType::InvalidScope => {
+                                warn!("OpenID Connect:  rfc-6749 5.2 {:?}", err);
+                                return Err(Error::ApiInsufficientRights(
+                                    "The Authorization was revoked for this user, client or action.".to_string(),
+                                ));
+                            }
+                            // The Extension Type Errors are used by the try_refresh_token method to signal generic problems with either the
+                            // current token, or the freshly received one.
+                            CoreErrorResponseType::Extension(err) => {
```

Comment:
According to an [Errata for RFC 6749](https://www.rfc-editor.org/errata/eid4745) two error codes are missing from section 5.2 of the RFC: `server_error` and `temporarily_unavailable`. However I see that the underlying `openidconnect` Rust create enum `BasicErrorResponseType` doesn't define these either, so I guess these will be presented to our client as `::Extension(str)`. It might be worth also noting that here.

Reply:
commit cf3fe15 deals with it in both try\_refresh\_token() and authenticate()